contracts/Otoken.sol

```diff
@@ -1,12 +1,173 @@
-/**
- * SPDX-License-Identifier: UNLICENSED
- */
-pragma solidity 0.6.10;
+/* SPDX-License-Identifier: UNLICENSED */
+pragma solidity =0.6.10;
+
+import {ERC20Initializable} from "./packages/oz/upgradeability/ERC20Initializable.sol";
+import {SafeMath} from "./packages/oz/SafeMath.sol";
+import {BokkyPooBahsDateTimeLibrary} from "./packages/BokkyPooBahsDateTimeLibrary.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
 
 /**
- *
+ * @title Otoken
+ * @author Opyn
+ * @notice Otoken is the ERC20 token for an option.
+ * @dev The Otoken inherits ERC20Initializable because we need to use the init instead of constructor.
  */
-// solhint-disable-next-line no-empty-blocks
-contract Otoken {
+contract Otoken is ERC20Initializable {
+    using SafeMath for uint256;
+    address public addressBook;
+    address public underlyingAsset;
+    address public strikeAsset;
+    address public collateralAsset;
+
+    uint256 public strikePrice;
+    uint256 public expiry;
+
+    bool public isPut;
+
+    uint256 private constant STRIKE_PRICE_DIGITS = 10**18;
+
+    constructor(address _addressBook) public {
+        addressBook = _addressBook;
+    }
+
+    /**
+     * @notice initialize a new otokens.
+     * @param _underlyingAsset asset that the option references
+     * @param _strikeAsset asset that the strike price is denominated in
+     * @param _collateralAsset asset that is held as collateral against short/written options
+     * @param _strikePrice strike price with decimals = 18
+     * @param _expiry expiration timestamp in second
+     * @param _isPut is this a put option, if not it is a call
+     */
+    function init(
+        address _underlyingAsset,
+        address _strikeAsset,
+        address _collateralAsset,
+        uint256 _strikePrice,
+        uint256 _expiry,
+        bool _isPut
+    ) external initializer {
+        underlyingAsset = _underlyingAsset;
+        strikeAsset = _strikeAsset;
+        collateralAsset = _collateralAsset;
+        strikePrice = _strikePrice;
+        expiry = _expiry;
+        isPut = _isPut;
+        (string memory name, string memory symbol) = _getNameAndSymbol(
+            _underlyingAsset,
+            _strikeAsset,
+            _collateralAsset,
+            _strikePrice,
+            _expiry,
+            _isPut
+        );
+        __ERC20_init_unchained(name, symbol);
+    }
+
+    /**
+     * @notice return option name as ETHUSDC/1597512453/200P/USDC, symbol ETHUSDC/1597512453/200P/USDC
+     * @return name ETH-USDC 200 Put 1597512453
+     * @return symbol ETHUSDC/1597512453/200P/USDC
```

Comment:
has this been decided or is this still TBD?

Reply:
TBD~